### PR TITLE
Add fix for JULIA_PKGDIR using relative path

### LINKED ISSUE #20896: JULIA_PKGDIR doesn't work with relative paths
```
├ JULIA_PKGDIR=jtest juliadev -e 'Pkg.init(); Pkg.add("Colors")'
INFO: Initializing package repository /Users/simon/jtest/v0.6
INFO: Cloning METADATA from https://github.com/JuliaLang/METADATA.jl
ERROR: SystemError: opening file /Users/simon/jtest/v0.6/jtest/v0.6/METADATA/ColorTypes/versions/0.3.4/sha1: No such file or directory
#systemerror#39 at ./error.jl:64 [inlined]
systemerror(::String, ::Bool) at ./error.jl:64
open(::String, ::Bool, ::Bool, ::Bool, ::Bool, ::Bool) at ./iostream.jl:104
open(::Base.#readstring, ::String) at ./iostream.jl:150
readstrip(::String, ::Vararg{String,N} where N) at ./pkg/read.jl:8
sha1(::String, ::VersionNumber) at ./pkg/read.jl:11
resolve(::Dict{String,Base.Pkg.Types.VersionSet}, ::Dict{String,Dict{VersionNumber,Base.Pkg.Types.Available}}, ::Dict{String,Tuple{VersionNumber,Bool}}, ::Dict{String,Base.Pkg.Types.Fixed}, ::Dict{String,VersionNumber}, ::Set{String}) at ./pkg/entry.jl:516
edit(::Function, ::String, ::Base.Pkg.Types.VersionSet, ::Vararg{Base.Pkg.Types.VersionSet,N} where N) at ./pkg/entry.jl:30
(::Base.Pkg.Entry.##1#3{String,Base.Pkg.Types.VersionSet})() at ./task.jl:335
Stacktrace:
 [1] sync_end() at ./task.jl:287
 [2] macro expansion at ./task.jl:303 [inlined]
 [3] add(::String, ::Base.Pkg.Types.VersionSet) at ./pkg/entry.jl:50
 [4] (::Base.Pkg.Dir.##2#3{Array{Any,1},Base.Pkg.Entry.#add,Tuple{String}})() at ./pkg/dir.jl:31
 [5] cd(::Base.Pkg.Dir.##2#3{Array{Any,1},Base.Pkg.Entry.#add,Tuple{String}}, ::String) at ./file.jl:69
 [6] #cd#1(::Array{Any,1}, ::Function, ::Function, ::String, ::Vararg{Any,N} where N) at ./pkg/dir.jl:31
 [7] add(::String) at ./pkg/pkg.jl:106
```

The problem seems to be that `Pkg.Dir.path` expects to be called from within the original working directory, whereas [here](https://github.com/JuliaLang/julia/blob/97c40edb05cd9a3d48e0633e9e1b40340f7f5ca5/base/pkg/read.jl#L11-L12) it is called from within `Pkg.cd`, which gives the incorrect result:
```
├ JULIA_PKGDIR=jtest juliadev -e 'println(Pkg.cd(Pkg.Dir.path,"METADATA"))'
/Users/simon/jtest/v0.6/jtest/v0.6/METADATA
```

### PR DESCRIPTION
**Pull request by simonbyrne**
Fixes #20896.

### CODE REVIEW

**Comment by StefanKarpinski** [comment]
I'm ok with this as a temporary solution, but the better solution would be what I described on the issue thread. Also unclear how much effort is worth putting into this given Pkg3's impendingness.

File: base/pkg/dir.jl [hunk 1]
```diff
@@ -28,7 +28,17 @@ function cd(f::Function, args...; kws...)
         !haskey(ENV,"JULIA_PKGDIR") ? init() :
             throw(PkgError("Package metadata directory $metadata_dir doesn't exist; run Pkg.init() to initialize it."))
     end
-    Base.cd(()->f(args...; kws...), dir)
+    if haskey(ENV,"JULIA_PKGDIR")
+        olddir = ENV["JULIA_PKGDIR"]
+        ENV["JULIA_PKGDIR"] = dir # use absolute path
```
Comment (tkelman) [inline base/pkg/dir.jl]: withenv?

Comment (StefanKarpinski) [inline base/pkg/dir.jl]: Yes, the do-block form of `withenv` already does all of this for you.

Comment (tkelman) [inline base/pkg/dir.jl]: and won't this append multiple copies of the version?

File: base/pkg/dir.jl [hunk 2]
```diff
@@ -29,12 +29,8 @@ function cd(f::Function, args...; kws...)
             throw(PkgError("Package metadata directory $metadata_dir doesn't exist; run Pkg.init() to initialize it."))
     end
     if haskey(ENV,"JULIA_PKGDIR")
-        olddir = ENV["JULIA_PKGDIR"]
-        ENV["JULIA_PKGDIR"] = dir # use absolute path
-        try
+        withenv("JULIA_PKGDIR" => _pkgroot()) do
```
Comment (tkelman) [inline base/pkg/dir.jl]: add a test that would have failed with double versions?

Comment (simonbyrne) [inline base/pkg/dir.jl]: I don't think the other one was actually giving double versions, as the `isdir(d) || !isdir(b)` test in `Dir.path()` would have failed. But this is clearer.

File: base/pkg/dir.jl [hunk 3]
```diff
@@ -28,7 +28,13 @@ function cd(f::Function, args...; kws...)
         !haskey(ENV,"JULIA_PKGDIR") ? init() :
             throw(PkgError("Package metadata directory $metadata_dir doesn't exist; run Pkg.init() to initialize it."))
     end
-    Base.cd(()->f(args...; kws...), dir)
+    if haskey(ENV,"JULIA_PKGDIR")
+        withenv("JULIA_PKGDIR" => _pkgroot()) do
```
Comment (tkelman) [inline base/pkg/dir.jl]: maybe this should just say `"JULIA_PKGDIR" => abspath(ENV["JULIA_PKGDIR"])` ? should be equivalent but takes less scrolling up to see what it's accomplishing

Comment (simonbyrne) [inline base/pkg/dir.jl]: seems reasonable.

**Pull request merged**